### DISCUSSION TITLE
Distinguishing Individual Objects in Exported Field Plot File

### DISCUSSION THREAD

**Discussion by chamanth-vct** [ansys/pyaedt#5334] [Q&A]
Hi 

I am currently using PyAEDT to export field results from AEDT and have encountered an issue. Below is a summary of what I’m doing and the problem I am facing:

### Process Overview:
To export field results, I use the following PyAEDT functionalities:

- `create_fieldplot_volume()` – Creates a field plot for specified objects.

- `export_field_plot()` – Exports the created field plot to a file.

One of the parameters in `create_fieldplot_volume()` is `assignment` (refer to [documentation link](https://aedt.docs.pyansys.com/version/stable/API/visualization/_autosummary/ansys.aedt.core.visualization.post.post_common_3d.PostProcessor3D.create_fieldplot_volume.html#ansys.aedt.core.visualization.post.post_common_3d.PostProcessor3D.create_fieldplot_volume:~:text=Parameters%3A-,assignment,-list)), where we specify a list of objects for which we want to generate the field results.

### Issue Description:

Single Object Export:

- When I pass a single object (e.g., 'coil1'), the exported file contains the field result specifically for that object.

Multiple Objects Export:

- When I pass multiple objects (e.g., 'coil1', 'coil2'), the exported file contains the combined field result for all the objects. However, the individual objects are not distinguishable in the exported file, making it challenging to identify their contributions separately.

I’ll further clarify the problem with the following example.

### Code Example

```py
import ansys.aedt.core 
from pathlib import Path
import pyvista as pv

folder_path = Path(r"C:\Users\chamanth\testing_error")
file_path =  folder_path.joinpath("VCollab_Example.aedt")     

exports_folder_path = folder_path.joinpath("exports_folder")
exports_folder_path.mkdir(parents=True,exist_ok=True)

desktop = ansys.aedt.core.Desktop() 
prj= desktop.load_project(file_path.__str__())
model = prj.modeler
part_names = model.object_names
post = prj.post  # post_processor

################# Single Object   #####################
object_list = ['coil1']

h_field_plot = post.create_fieldplot_volume(object_list,quantity="H_Vector",
                                                plot_name="H_vector_plot")

result_file_name = exports_folder_path.joinpath('single_object')
export_status = post.export_field_plot(plot_name = "H_vector_plot",
                        output_dir = exports_folder_path,
                        file_name = result_file_name.__str__(),
                        file_format = 'case')

h_field_plot.delete()

################# Multiple Object #####################
object_list = ['coil1','coil2']

h_field_plot = post.create_fieldplot_volume(object_list,quantity="H_Vector",
                                                plot_name="H_vector_plot")

result_file_name = exports_folder_path.joinpath('multi_object')
export_status = post.export_field_plot(plot_name = "H_vector_plot",
                        output_dir = exports_folder_path,
                        file_name = result_file_name.__str__(),
                        file_format = 'case')

h_field_plot.delete()

desktop.close_desktop()
```
In both cases (single or multiple objects), the exported .case files contain a single mesh. When multiple objects are assigned, I can’t distinguish individual objects within the exported file.

Download the sample file used in the code from this [link](https://apps.vcollab.com/fileshare/downloadsharedfile?file_ID=3ob590m2lpchjb).

### Challenges:
At present, to get the field result for each individual object, I export the results separately for each object.

This approach works fine for small models with a few objects. However, it becomes _**time-consuming and inefficient for larger models with many parts.**_

### Question:

I am looking for a way to optimize the export process by reducing the time taken for large models while retaining object-level information in the exported .case file.

Is there any feature or workaround that would allow me to export results for multiple objects at once while still distinguishing the individual objects in the output?

Any suggestions or guidance would be greatly appreciated!

Thank you!

**Comment by gmalinve** [comment]
Hi @chamanth-vct,

Thanks for opening this discussion.

I need to understand something first:
- Which application are you using (Maxwell, HFSS, Q3D etc.)?
- Which solver are you using?
- Is this discussion related to this #5317? if so, can you give us some context about your doubt on the missing surfaces in the mesh plots?

Thanks,

Kind regards,

Giulia 

@tizianrot

**Comment by chamanth-vct** [comment]
Hi @gmalinve 

Thank you for your response.
I am using both Maxwell and HFSS.
This discussion is not exactly related to the issue you mentioned. Here's the distinction:

**This Discussion:**

- When I create a field plot for multiple objects and export it as a .case file, I am unable to identify the individual objects within the exported file. The field data in the exported file appears combined, making it difficult to distinguish between the contributions of individual objects.


**The Issue in #5317:**

- In that issue, I generate both a mesh plot and a field plot for a single object and export them into individual .case files. The problem arises when I visualize both exported .case files – there is a mismatch between the mesh plot and the field plot, as described in that issue.

I hope this clarifies the distinction between the two topics. Let me know if you need further details.

**Comment by lorenzovecchietti** [comment]
Hi @chamanth-vct, based on what I know, the native APIs don’t seem to support what you're aiming to do. If your end goal is to work with PyVista, one option could be to split your field using the geometry exported from export_model_obj.

Could you provide more details on what you're trying to accomplish? With a better understanding, we might be able to suggest alternative approaches.

**Comment by gmalinve** [comment]
@chamanth-vct I have an additional question, since PyAEDT supports the export of plots in two formats (.aedtplt and .case) and can also leverage PyVista for post-processing visualization, is there a specific reason for your to export in .case? is your goal to work with Ensight?

**Comment by chamanth-vct** [comment]
Hi @lorenzovecchietti and @gmalinve,

Thank you for your insights and questions.

**Goal:**

My aim is to access the mesh and field result data of the objects (parts) within AEDT files for post-processing and further analysis using our custom scripts. Since I haven’t found a direct method in PyAEDT to retrieve this data, I’m currently exporting mesh and field results to use them downstream.



**Export Format and Limitations:**

I'm open to using either `.aedtplt` or `.case` formats, as both work within our workflow. However, I’ve encountered the same limitation in both: when exporting field data for multiple objects, it’s not possible to identify or separate individual parts within the resulting file. This poses a challenge in isolating data for specific objects during post-processing.

If there’s an approach that allows accessing mesh and field result data at the object level, I would be very interested to know more.

I hope this clarifies my objective. Please let me know if you have further questions or need additional details.

Thanks again for your support!

**Comment by lorenzovecchietti** [comment]
You cannot do this neither with pyAEDT neither with native APIs. The only option is exporting each object one by one. I agree this is not ideal and I opened an enhancement request.

Do you need separated objects just for plotting purposes, or for something else?

If it is for plotting purposes, I think that the only option is to work with pyVista functions and set a workflow like this:
1. export all the objects individually using `export_model_obj`
2. export the whole fields (without any geometry separation)
3. use pyVista functions to split the fields, using the objs as a filter ([boolean operations](https://docs.pyvista.org/examples/01-filter/boolean-operations)?)

This is just an idea I never tried myself, if you try it, let me know if it works!

If you have a need other than plotting, please let us know, and maybe we can find some other workflows to suggest.
